### ISSUE TITLE
What is the cannonical way of using monaco.languages.registerCompletionItemProvider in the React portion of the app?

### ISSUE THREAD

**Issue by benjamin-weller** [react-monaco-editor/react-monaco-editor#316]
I've seen a few questions like #249 and #88. I have checked your example implementation at the beginning of the project, but frankly that `registerCompletionItemProvider` was created at the beginning of the project, and not in the react portion.

I've personally copy pasted the default [monaco playground completion provider example](https://microsoft.github.io/monaco-editor/playground.html#extending-language-services-completion-provider-example) and placed the `registerCompletionItemProvider` in `editorDidMount`, but to no avail.

It would appear that a variety of folks would really be able to use a example implementation of `registerCompletionItemProvider` in the React portion of the app, as it doesn't currently match what does work just using the standard monaco library.

**Comment by codebykat** [comment]
In case it helps anyone else, this is working for us in `editorDidMount`, example code here: https://github.com/Automattic/simplenote-electron/blob/06edf9dc27d71ffeb6dc763ffe55a7c87c31b85a/lib/note-content-editor.tsx#L196, https://github.com/Automattic/simplenote-electron/blob/06edf9dc27d71ffeb6dc763ffe55a7c87c31b85a/lib/note-content-editor.tsx#L822

**Comment by mike-lischke** [comment]
If you guys had asked that on Stackoverflow, it would be more fun to answer, but anyway:

All providers are a per language, but not per editor setting. So it makes no sense to set providers in a mount action (btw. I never used `editorDidMount`etc. Instead everything works fine with the standard React functions (`componentDidMount` etc.).

For initialisation I use a static function:

```
    private static configureMonaco(): void {
        const completionProvider = new CodeCompletionProvider();
        const hoverProvider = new HoverProvider();
        const signatureHelpProvider = new SignatureHelpProvider();
        const documentHighlightProvider = new DocumentHighlightProvider();
        const definitionProvider = new DefinitionProvider();
        const referenceProvider = new ReferencesProvider();
        const formattingProvider = new FormattingProvider();
        const renameProvider = new RenameProvider();

        for (const language of ["mysql", "sql", "python"]) {
            languages.registerCompletionItemProvider(language, completionProvider);
            languages.registerHoverProvider(language, hoverProvider);
            languages.registerSignatureHelpProvider(language, signatureHelpProvider);
            languages.registerDocumentHighlightProvider(language, documentHighlightProvider);
            languages.registerDefinitionProvider(language, definitionProvider);
            languages.registerReferenceProvider(language, referenceProvider);
            languages.registerDocumentFormattingEditProvider(language, formattingProvider);
            languages.registerRenameProvider(language, renameProvider);
        }

        if (languages.typescript) { // This field is not set when running under Jest.
            languages.typescript.javascriptDefaults.setCompilerOptions({
                allowNonTsExtensions: true,
                target: languages.typescript.ScriptTarget.ES2017,
                lib: ["es2017"],
            });

            languages.typescript.typescriptDefaults.setCompilerOptions({
                allowNonTsExtensions: true,
                target: languages.typescript.ScriptTarget.ES2017,
                lib: ["es2017"],
            });
        }
    }
```

Originally I ran this function on startup of my react app, but since that didn't work well with Jest I now load that when I instantiate my first code editor (in the constructor) and the init wasn't done yet (there's a flag).

**Comment by sivasini** [comment]
@csurfleet Hi, I am stuck at the same situation presently. Did you arrive at any solution regarding the problem?

**Comment by Muffinmann** [comment]
For those who counter the problem that registerCompletionItemProvider doesn't work:

**TL;DR**: make sure the `suggest` module is included correctly in your project. In my case(vite-react), I have to do:
`import  'monaco-editor/esm/vs/editor/editor.all.js`  

**The reason**
I followed [this sample](https://github.com/microsoft/monaco-editor/tree/main/samples/browser-esm-vite-react) to set up my project, and ran into this issue. After checking the source code I realized that the `suggest` module is not bundled into my app:
![image](https://github.com/react-monaco-editor/react-monaco-editor/assets/24995804/4b339999-fa09-4fb2-8158-450e7375d7d9)
In the `package.json` file of monaco-editor you can see [the entry point](https://github.com/microsoft/monaco-editor/blob/a5e379c985d3263aacb1e811958c4663fea3db67/package.json#L34) and it doesn't include files from the `contrib` folder. This means `import * as monaco from 'monaco-editor'` does not really import the `contrib` folder where the `suggest` module is located.

I hope this could be helpful and please let me know if anything of my explanation is incorrect.
